### PR TITLE
TestEngine: add Checkpoints, RpcStorage and Dump to html

### PR DESCRIPTION
**Pull request by shargon**
Part of https://github.com/neo-project/neo-devpack-dotnet/issues/889

Features added:
- Checkpoints
- RpcStorage
- Dump coverage to html
image
- Changed coverage reports to show the arg name
```
0xf429761491dff75d93d8c6aa466b1aab1eaa0950 [100.00%]
┌-─────────────────────────────-┬-───────-┐
│ Method                        │  Line   │
├-─────────────────────────────-┼-───────-┤
│ _deploy(data,update)          │ 100.00% │
│ _initialize()                 │ 100.00% │
│ balanceOf(owner)              │ 100.00% │
│ burn(account,amount)          │ 100.00% │
│ decimals()                    │ 100.00% │
│ getOwner()                    │ 100.00% │
│ mint(to,amount)               │ 100.00% │
│ myMethod()                    │ 100.00% │
│ setOwner(newOwner)            │ 100.00% │
│ symbol()                      │ 100.00% │
│ totalSupply()                 │ 100.00% │
│ transfer(from,to,amount,data) │ 100.00% │
│ update(nefFile,manifest,data) │ 100.00% │
│ verify()                      │ 100.00% │
└-─────────────────────────────-┴-───────-┘
```

### CODE REVIEW

**Comment by Jim8y** [comment]
Nice feature

**Comment by Jim8y** [comment]
Any plan to udpate the readme as well?

File: src/Neo.SmartContract.Testing/Storage/EngineStorage.cs [hunk 1]
```diff
@@ -33,7 +33,7 @@ public class TestStorage
         /// Constructor
         /// </summary>
         /// <param name="storage">Store</param>
```
Comment (Jim8y) [inline src/Neo.SmartContract.Testing/Storage/EngineStorage.cs:35]:
```suggestion
        /// <param name="store">Store</param>
```

**Comment by Jim8y** [comment]
Not a problem here, but in nativeartifact:

```
         /// <summary>
        /// OracleContract
        /// </StdLib>
        public StdLib StdLib { get; }
```

**Comment by shargon** [comment]
> StdLib

Fixed in nep17 template to avoid conflicts

**Comment by shargon** [comment]
`RpcStorage` is `Seek` limited, we can't seek `Backward` => `https://github.com/neo-project/neo-modules/blob/7faf5356ce86e49e10a75fcffa0aba654f16cd9d/src/RpcServer/RpcServer.Blockchain.cs#L234

File: tests/Neo.SmartContract.Testing.UnitTests/Storage/TestStorageTests.cs [hunk 2]
```diff
@@ -1,19 +1,47 @@
...
+            var engine = new TestEngine(true);
+
+            Assert.AreEqual(100_000_000, engine.Native.NEO.TotalSupply);
+
+            var checkpoint = engine.Storage.Checkpoint();
+
+            // Test restoring the checkpoint
+
+            var storage = new EngineStorage(new MemoryStore());
+            checkpoint.Restore(storage.Snapshot);
```
Comment (Jim8y) [inline tests/Neo.SmartContract.Testing.UnitTests/Storage/TestStorageTests.cs:27]: Just curious here, should not it be storage.restore(checkpoint) make more sense?

File: tests/Neo.SmartContract.Testing.UnitTests/Storage/Rpc/RpcStoreTests.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            public abstract BigInteger GetCandidateVote(ECPoint point);
+            protected DummyContract(SmartContractInitialize initialize) : base(initialize) { }
+        }
+
+        [TestMethod]
+        public void TestRpcStore()
+        {
+            var engine = new TestEngine(false)
+            {
+                Storage = new EngineStorage(new RpcStore("http://seed2.neo.org:10332"))
```
Comment (Jim8y) [inline tests/Neo.SmartContract.Testing.UnitTests/Storage/Rpc/RpcStoreTests.cs:24]: Test net? or mock rpc?

Comment (shargon) [reply tests/Neo.SmartContract.Testing.UnitTests/Storage/Rpc/RpcStoreTests.cs:24]: Real rpc

Comment (Jim8y) [reply tests/Neo.SmartContract.Testing.UnitTests/Storage/Rpc/RpcStoreTests.cs:24]: I mean, should it not be a testnet or better mock rpc? I am cool if you insist with real rpc. We can update later.

**Review by Jim8y — DISMISSED** [review]

File: tests/Neo.SmartContract.Testing.UnitTests/Storage/TestStorageTests.cs (hunk 2 above)
Comment (shargon) [reply tests/Neo.SmartContract.Testing.UnitTests/Storage/TestStorageTests.cs:27]: We can do both of them

File: tests/Neo.SmartContract.Testing.UnitTests/Storage/Rpc/RpcStoreTests.cs [hunk 4]
```diff
@@ -29,9 +29,9 @@ public void TestRpcStore()
             Assert.AreEqual(100_000_000, engine.Native.NEO.TotalSupply);
             Assert.IsTrue(engine.Native.Ledger.CurrentIndex > 4_905_187);
 
-            // check with Seek (RPC doesn't support Backward)
+            // check with Seek (RPC doesn't support Backward, it could be slow)
 
-            // Assert.IsTrue(engine.Native.NEO.GasPerBlock == 5);
+            Assert.IsTrue(engine.Native.NEO.GasPerBlock == 500000000);
```
Comment (Jim8y) [inline tests/Neo.SmartContract.Testing.UnitTests/Storage/Rpc/RpcStoreTests.cs:34]: LOL, i did not notice this at all, see, we do need `datoshi`

File: src/Neo.SmartContract.Testing/Storage/Rpc/RpcStore.cs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    public ISnapshot GetSnapshot() => new RpcSnapshot(this);
+    public bool Contains(byte[] key) => TryGet(key) != null;
+    public void Dispose() { }
+
+    #region Rpc calls
+
+    public IEnumerable<(byte[] Key, byte[] Value)> Seek(byte[] key, SeekDirection direction)
+    {
+        if (direction is SeekDirection.Backward)
+        {
```
Comment (shargon) [inline src/Neo.SmartContract.Testing/Storage/Rpc/RpcStore.cs:51]: @Jim8y it should be great to have this in rpc

**Comment by shargon** [comment]
I will merge it after https://github.com/neo-project/neo-devpack-dotnet/pull/893 , I prefer try to avoid conflicts there, meanwhile I can write the readme here

**Comment by shargon** [comment]
image

New feature

**Comment by shargon** [comment]
```
 0x6ba9a0996663b25d54d3c293e18232be26df3a70 [97.37%]
 ┌-─────────────-┬-───────-┐
 │ Method        │  Line   │
 ├-─────────────-┼-───────-┤
 │ _initialize,0 │ 100.00% │
 │ balanceOf,1   │ 100.00% │
 │ burn,2        │ 100.00% │
 │ decimals,0    │ 100.00% │
 │ mint,2        │ 100.00% │
 │ myMethod,0    │ 100.00% │
 │ setOwner,1    │ 100.00% │
 │ symbol,0      │ 100.00% │
 │ totalSupply,0 │ 100.00% │
 │ update,3      │ 100.00% │
 │ verify,0      │ 100.00% │
 │ getOwner,0    │  95.24% │
 │ transfer,4    │  94.62% │
 │ _deploy,2     │  94.44% │
 └-─────────────-┴-───────-┘
 ```

**Comment by shargon** [comment]
Ready to review @cschuchardt88 @Jim8y , 100% coverage raised! 🚀

File: src/Neo.SmartContract.Testing/TestingApplicationEngine.cs [hunk 6]
```diff
@@ -42,10 +45,24 @@ protected override void PreExecuteInstruction(Instruction instruction)
             base.PreExecuteInstruction(instruction);
         }
 
+        protected override void OnFault(Exception ex)
+        {
+            base.OnFault(ex);
+
+            if (PreInstruction is not null)
+            {
```
Comment (shargon) [inline src/Neo.SmartContract.Testing/TestingApplicationEngine.cs:53]:
```suggestion
            {
            // PostExecuteInstruction is not executed onFault
```

File: tests/Neo.SmartContract.Template.UnitTests/templates/neocontractnep17/CoverageContractTests.cs [hunk 7]
```diff
@@ -8,20 +8,23 @@ public class CoverageContractTests
...
         public static void EnsureCoverage()
         {
-            // Ennsure that the coverage is more than X% at the end of the tests
+            // Join here all of your Coverage sources
 
             var coverage = Nep17ContractTests.Coverage;
             coverage?.Join(OwnerContractTests.Coverage);
 
-            Assert.IsNotNull(coverage);
+            // Ennsure that the coverage is more than X% at the end of the tests
```
Comment (shargon) [inline tests/Neo.SmartContract.Template.UnitTests/templates/neocontractnep17/CoverageContractTests.cs:21]:
```suggestion
            // Ensure that the coverage is more than X% at the end of the tests
```

**Comment by shargon** [comment]
@Jim8y merge? it can be optimized in a different pr

**Review by Jim8y — APPROVED** [review]

**Pull request merged**